content/tutorials/build-apps/build-a-desktop-wallet-in-javascript.md

```diff
@@ -0,0 +1,1422 @@
+
+```html
+<body>
+    <!-- Step 1 code modifications - start -->
+    <h3>Build a XRPL Wallet</h3>
+    Latest validated ledger index: <strong id="ledger-index"></strong>
+    <!-- Step 1 code modifications - end -->
+</body>
+```
+
+5. In `view/template.html` add the following line at the bottom of the file:
+
+```html
+</body>
+<!-- Step 1 code additions - start -->
+<script src="renderer.js"></script>
+<!-- Step 1 code additions - end -->
+</html>
+```
+
+6. Now create the `renderer.js` file in the `view`folder with the following code:
+
+```javascript
+const ledgerIndexEl = document.getElementById('ledger-index')
+
+// Here we define the callback function that performs the content update
+// whenever 'update-ledger-index' is called by the main process
+window.electronAPI.onUpdateLedgerIndex((_event, value) => {
+    ledgerIndexEl.innerText = value
+})
+```
+
+7. To wire up our main application to send the ledger index to the frontend, modify `index.js` by adding the following snippet:
+
+```javascript
+// Here we have to wait for the application to signal that it is ready
+// to execute our code. In this case we create a main window, query
+// the ledger for its latest index and submit the result to the main
+// window where it will be displayed
+app.whenReady().then(() => {
+  // Step 1 code additions - start
+  const appWindow = createWindow()
+  
+  getValidatedLedgerIndex().then((value) => {
+    appWindow.webContents.send('update-ledger-index', value)
+  })
+  // Step 2 code additions - end
+  
+})
+```
+
+Here we first call our helper function `getValidatedLedgerIndex()` and then broadcast an event named `update-ledger-index` and an attached payload which can be handled by the frontend. 
+
+This example shows how to do Inter Process Communication (IPC) in Electron. Technically, JavaScript has no true parallel processes or threading, because it follows a single-threaded event-driven paradigm. Nonetheless Electron provides us with two IPC modules called `ipcMain` and `ipcRenderer`. We can roughly equate those two to a backend process and a frontend process when we think in terms of client-server applications. It works as follows:
+
+1. We started by creating a function that enables the frontend to subscribe to backend events via the ContextBridge (in `view/preload.js`)
+2. Then we make the function available by putting it in a preloader script to ensure it is loaded and can be used by the frontend.
+3. On the frontend, we can then use that function to attach a callback that handles frontend updates when the event is dispatched. We could do this in the console, in a `<script>` tag in the frontend or in our case in a separate file.
+5. Lastly, we dispatch the event from the main logic in `index.js`via `appWindow.webContents.send('update-ledger-index', value)`.
+
+To get the application running at this early stage of development, run the following command:
+
+```console
+npm run start
+```
```

Comment:
Could you add an expected value for each of these "final" steps so they know what to look for? (Downloading & running the reference code seems like a lot of extra steps to verify they're doing ok, when in practice just knowing one thing to look for is enough. 

Ex. "You should see something like 'Latest validated ledger index: 39296259' the number will be different since it is the latest ledger index as of when the app started. We'll make it continuously update later on."

Reply:
Good idea, i did that for every step. Gonna close it  with that.